rust/sealevel/programs/interchain-accounts/src/processor.rs

```diff
@@ -0,0 +1,507 @@
+#[macro_export]
+macro_rules! call_remote_authority_pda_seeds {
+    () => {{
+        &[
+            b"hyperlane_interchain_account",
+            b"-",
+            b"call_remote_authority",
+        ]
+    }};
+
+    ($bump_seed:expr) => {{
+        &[
+            b"hyperlane_interchain_account",
+            b"-",
+            b"call_remote_authority",
+            &[$bump_seed],
+        ]
+    }};
+}
+
+#[cfg(not(feature = "no-entrypoint"))]
+solana_program::entrypoint!(process_instruction);
+
+/// The program's entrypoint.
+pub fn process_instruction(
+    program_id: &Pubkey,
+    accounts: &[AccountInfo],
+    instruction_data: &[u8],
+) -> ProgramResult {
+    if let Ok(recipient_instruction) = MessageRecipientInstruction::decode(instruction_data) {
+        return match recipient_instruction {
+            MessageRecipientInstruction::InterchainSecurityModule => {
+                get_interchain_security_module(program_id, accounts)
+            }
+            MessageRecipientInstruction::InterchainSecurityModuleAccountMetas => {
+                set_account_meta_return_data(program_id)
+            }
+            MessageRecipientInstruction::Handle(_instruction) => {
+                unimplemented!("Handle instruction not implemented")
+            }
+            MessageRecipientInstruction::HandleAccountMetas(_) => {
+                unimplemented!("HandleAccountMetas instruction not implemented")
+            }
+        };
+    }
+
+    let instruction = InterchainAccountInstruction::try_from_slice(instruction_data)
+        .map_err(|_| ProgramError::InvalidInstructionData)?;
+    match instruction {
+        InterchainAccountInstruction::Init(mailbox) => init(program_id, accounts, mailbox),
+        InterchainAccountInstruction::CallRemote(remote_call) => {
+            send_call_remote(program_id, accounts, remote_call)
+        }
+        InterchainAccountInstruction::SetInterchainSecurityModule(ism) => {
+            set_interchain_security_module(program_id, accounts, ism)
+        }
+        InterchainAccountInstruction::EnrollRemoteRouters(configs) => {
+            enroll_remote_routers(program_id, accounts, configs)
+        }
+    }
+}
+
+/// Creates the storage PDA.
+///
+/// Accounts:
+/// 0. `[executable]` System program.
+/// 1. `[signer]` Payer.
+/// 2. `[writeable]` Storage PDA.
+fn init(program_id: &Pubkey, accounts: &[AccountInfo], init: Init) -> ProgramResult {
+    let accounts_iter = &mut accounts.iter();
+
+    // Account 0: System program.
+    let system_program_info = next_account_info(accounts_iter)?;
+    if system_program_info.key != &system_program::id() {
+        return Err(ProgramError::InvalidArgument);
+    }
+
+    // Account 1: Payer.
+    let payer_info = next_account_info(accounts_iter)?;
+    if !payer_info.is_signer {
+        return Err(ProgramError::MissingRequiredSignature);
+    }
+
+    // Account 2: Storage PDA.
+    let storage_info = next_account_info(accounts_iter)?;
```

Comment:
believe we'll need to make sure that the program hasn't been initialized before (fwiw I wouldn't be surprised if helloworld doesn't have the correct safety checks, it was meant more for testing than anything)

Take a look at `verify_account_uninitialized` which we can use on the storage_info